Fix Game Interface / AC Callbacks + support for T5ZM, T6MP and T6ZM for the Game Interface.

xerxes-at opened the pull request:
Fixes the following:

* On IW4, IW5 and T5 it was possible that `level.eventBus.gamename` was being set before `level.eventBus` was initialized as both happened through the init method in 2 different gsc files.

* Paths in `_integration_shared.gsc` pointing to `_integration_base.gsc` did not account for the different paths caused by deploy.bat moving the files to mp only folders for IW5 and T5.

* T6's _customcallbacks.gsc did not compile with the new compiler.

Adds the following:

* Support for T5ZM
    * Remove mp includes from `_integration_base`.
    * Define `GetXuid` as `overrideMethod` as T5ZM doesn't have it.
    * Define `GetPlayerFromClientNum` as `overrideMethod`  since getting all players is slightly different on T5ZM.
 * Support for T6 (MP and ZM)
    * Add `_integration_t6zm_helper.gsc` to support the same hud messages as MP.
    * Add `_integration_t6.gsc`

Other changes:

* Remove precompiled `_customcallbacks.gsc` for T6 as it can now load uncompiled GSC and the compiler used by Pluto is much better than the old one.
* Changes to deploy.bat to support the other changes.

xerxes-at committed:
Fix trying to write to a struct before its initialized.

Same issue on IW4, IW5 and T5 game modules.

xerxes-at committed:
Fix path issues in the scripts + add support for t5zm.

* Fix deploy.bat
* Change paths inside the gsc scripts used to call functions in other scripts
* Remove mp includes from base gsc file.
#include maps\mp\_utility;
#include maps\mp\gametypes\_hud_util;
* Define GetXuid as overrideMethod as t5zm doesn't have it.
* Define GetPlayerFromClientNum as getting all players is slightly different on t5zm.

xerxes-at committed:
Remove the precompiled gsc file for T6 as PlutoT6 can load uncompiled GSC now.

xerxes-at committed:
Fix _customcallbacks.gsc for T6

xerxes-at committed:
Add T6 support to the game interface.

diamante0018 reviewed (commented):
You should removed the compiled T6 plugin. There should not be a reason to encourage loading scripts that way.

diamante0018 commented on GameFiles/GameInterface/_integration_iw4x.gsc:
```diff
@@ -199,6 +199,11 @@ Log2Console( logLevel, message )
     PrintConsole( "[" + logLevel + "] " + message + "\n" );
 }
 
+_GetXUID()
```
No this is wrong. See how `ISTestClient` function is called in the same file.
You need to use the `call` keyword.

diamante0018 reviewed (changes requested):
You should removed the compiled T6 plugin. There should not be a reason t encourage loading scripts that way.

diamante0018 commented on GameFiles/GameInterface/_integration_iw5.gsc:
```diff
@@ -109,6 +109,11 @@ Log2Console( logLevel, message )
     Print( "[" + logLevel + "] " + message + "\n" );
 }
 
+_GetXUID()
```
Same as iw4 engine

diamante0018 commented on GameFiles/GameInterface/_integration_t6.gsc:
```diff
@@ -0,0 +1,569 @@
+
+    self IPrintLnBold( "NoClip enabled" );*/
+
+    scripts\_integration_base::LogWarning( "NoClip is not supported on T6!" );
```
Noclip should be a built-in on t6 iirc

diamante0018 commented on GameFiles/GameInterface/_integration_t6.gsc:
```diff
@@ -0,0 +1,569 @@
+        level waittill( "connected", player );
+        
+        if ( scripts\_integration_base::_IsBot( player ) ) 
+        {
```
Use istestclient as it is a built-in

diamante0018 commented on GameFiles/GameInterface/_integration_t5zm.gsc:
```diff
@@ -0,0 +1,550 @@
+        level waittill( "connected", player );
+        
+        if ( scripts\_integration_base::_IsBot( player ) ) 
+        {
```
Same

diamante0018 commented on GameFiles/GameInterface/_integration_t5.gsc:
```diff
@@ -129,6 +129,11 @@ God()
 }
 
+_GetXUID()
+{
```
On the 3arc engine you do not need the call keyword to call a built-in by reference but it should be attempted anyway

diamante0018 reviewed (approved):
Most of my comments can be discarded.
At the end because of the 3arc engine the call keyword can't be used so it forces us to use wrapper function.
I confirmed noclip is not supported on t5/t6 yet.
but I could be added to iw5 if it's not already done.

RaidMax commented on GameFiles/GameInterface/_integration_base.gsc:
```diff
@@ -22,6 +20,8 @@ Setup()
     level.commonFunctions.setDvar      = "SetDvarIfUninitialized";
     level.commonFunctions.isBot        = "IsBot";
+    level.commonFunctions.getXuid      = "GetXuid";
+    level.commonFunctions.GetPlayerFromClientNum = "GetPlayerFromClientNum";
```
nit: use camelCase for functionName
`level.commonFunctions.getPlayerFromClientNum`

RaidMax commented on GameFiles/GameInterface/_integration_base.gsc:
```diff
@@ -360,7 +361,7 @@ GenerateJoinTeamString( isSpectator )
         }
     }
 
-    guid = self GetXuid();
+    guid = self [[level.overrideMethods[level.commonFunctions.getXuid]]]();
```
So after thinking about this, I think we should move the `GenerateJoinTeamString` into the _integration_shared.gsc 
I added this in here before I created the shared functionality file.
Philosophically the _integration_base,gsc should really only be responsible for facilitating the game interface data transfer/notifications. No direct game interaction or action on game events.

That should put the above registration of the overrideMethods more in line with what's already done in the shared gsc

RaidMax commented:
What's the reason for defining an entirely new set of methods for T5ZM? Could we not check a dvar to see if it's zm vs mp and then register the GetXuid and GetPlayerFromClientNum respectively?

Edit: Actually I just realized it might complain about referencing the method even if you aren't calling it.. Which if so is annoying.

xerxes-at committed:
Update _integration_base.gsc

use camelCase for functionName

xerxes-at committed:
Make sure the Setup functions are always called in the right order.

Base -> shared -> game
Otherwise we might write to structs before they are created.

xerxes-at committed:
Move functions interacting with the game from _base to _shared

GetPlayerFromClientNum
OnPlayerJoinedTeam
OnPlayerJoinedSpectators
GenerateJoinTeamString
PlayerTrackingOnInterval
SaveTrackingMetrics

xerxes-at committed:
Block execution until game specific setup is done

Block _shared execution until the game specific file finished.
This allows the game specific file to override the events in _shared.

xerxes-at committed:
Fix setup event flow

Move check of sv_iw4madmin_integration_enabled dvar after waittill in _shared so _base has a chance to set it to 1.
Move check of sv_iw4madmin_autobalance dvar to OnPlayerConnect in _shared so the game specific script has a chance to set the dvar.

xerxes-at committed:
ignore bots

diamante0018 commented on GameFiles/GameInterface/_integration_t6.gsc:
```diff
@@ -23,6 +23,8 @@ Setup()
     
     RegisterClientCommands();
     
+    _SetDvarIfUninitialized("sv_iw4madmin_autobalance",0);
```
minor complaint here but could you leave a space after the comma and also space out the parameters like in the rest of the file?

xerxes-at commented:
Your guess regarding t5zm is dead on, unlike on t6 it will complain about any and all mp scripts.

xerxes-at committed:
add more spaces
